Deduplicate identical inbox color schemes to heal user settings

DHowett opened the pull request:
Up until now, we have treated inbox, fragment and user color schemes the
same: we load them all into one big map and when we save the settings
file we write them *all* out. It's been a big annoyance pretty much
forever.

In addition to cluttering the user's settings file, it prevents us from
making changes to the stock color schemes (like to change the cursor
color, or to adjust the colors in Tango Dark, or what have you) because
they're already copied in full in the user settings. It also means that
we need some special UI affordances for color schemes that you are
allowed to view but not to delete or rename.

We also have a funny hardcoded list of color scheme names and we use
that to determine whether they're "inbox" for UI purposes.

Because of all that, we are hesitant to add *more* color schemes to the
default set.

This pull request resolves all of those issues at once.

It:
- Adds an "origin" to color schemes indicating where they're from
  (Inbox, Fragment, User, ...)
- Replaces the Edit UI with a much simpler version that pretty much only
  has a "duplicate this color scheme to start editing it" button
- Deletes color schemes that we consider to be equivalent to inbox ones;
  this allows us to finally disentangle the user's preferences from the
  terminal's.
- Migrates all user settings that referred to schemes they may have
  modified (even implicitly!) to their modified versions.

The equivalence check intentionally leaves out the cursor and selection
colors, so that we have the freedom to change them in the future.

The Origin is part of a new interface, `ISettingsModelObject`, which we
can use in the future for things like Themes and Actions.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -991,14 +1051,14 @@ Json::Value CascadiaSettings::ToJson() const
     {
         const auto scheme{ winrt::get_self<ColorScheme>(entry.Value()) };
-        schemes.append(scheme->ToJson());
+        if (scheme->Origin() == OriginTag::User)
```
is this bodgy i can't tell

DHowett commented:
I'll include this in Leonard's new "Fix User Settings" code :)

zadjii-msft commented:
Notes from team sync today (post hackathon)

* The SUI still needs updates
* Instead of making the builtin schemes read-only, just sneaky save them as the `(Modified)` version

DHowett commented:
I am noodling on what the UI looks like when you try to edit one of these. We don't need to fun disclaimer text any more (woo), but we do need a call to action.

![image](https://github.com/microsoft/terminal/assets/189190/22c40d6f-5078-46ca-9a07-e391d519ec04)

zadjii-msft commented:
I'm gonna say :shipit:. I'd maybe make it an InfoBar but that's basically the same UX

DHowett commented:
Okay, this works pretty well.

![image](https://drop.howett.net/share/WindowsTerminal_20240223T130335_267.gif)

zadjii-msft commented:
okay mega :shipit:

zadjii-msft commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -642,7 +687,11 @@ void SettingsLoader::_parseFragment(const winrt::hstring& source, const std::str
                 {
-                    settings.globals->AddColorScheme(*scheme);
+                    scheme->Origin(OriginTag::Fragment);
+		    // Don't add the color scheme to the Fragment's GlobalSettings; that will
```
tabs?

zadjii-msft commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.h:
```diff
@@ -44,6 +44,8 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         winrt::com_ptr<implementation::Profile> baseLayerProfile;
         std::vector<winrt::com_ptr<implementation::Profile>> profiles;
         std::unordered_map<winrt::guid, winrt::com_ptr<implementation::Profile>> profilesByGuid;
+        std::unordered_map<winrt::hstring, winrt::com_ptr<implementation::ColorScheme>> colorSchemes;
```
oh wait no this is in `ParsedSettings` k this makes sense.

lhecker commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -403,6 +411,39 @@ bool SettingsLoader::DisableDeletedProfiles()
+        }
+    };
+
+    remapForAppearance(profile->DefaultAppearance());
+
+    if (auto unfocusedAppearance{ profile->UnfocusedAppearance() })
+    {
+        remapForAppearance(unfocusedAppearance);
+    }
```
An alternative way to express this would be something like this:
```cpp
const IAppearanceConfig appearances[] = {
    profile->DefaultAppearance(),
    profile->UnfocusedAppearance(),
};

for (const auto& appearance : appearances) {
    if (!appearance) {
        continue;
    }

    // the lambda contents
}
```

DHowett replied:
I prefer this, thanks

lhecker commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -796,6 +853,33 @@ void SettingsLoader::_addUserProfileParent(const winrt::com_ptr<implementation::
+            it->second = newScheme; // Stomp the user's existing scheme with the one we just got (to make sure the right Origin is set)
+            if (!existingScheme->IsEquivalentForSettingsMergePurposes(newScheme))
+            {
+                hstring newName{ fmt::format(L"{} (modified)", existingScheme->Name()) };
+                int differentiator = 2;
+                while (userSettings.colorSchemes.contains(newName))
+                {
+                    newName = hstring{ fmt::format(L"{} (modified {})", existingScheme->Name(), differentiator++) };
+                }
```
Shouldn't this result in `fixup == true`?

lhecker replied:
(This is my only question before approval btw. The others ones are nits.)

DHowett replied:
Great catch! I've added a test for this and yes, it fails :)

lhecker reviewed (commented):
The `fmt::format` calls throughout the PR could use `FMT_COMPILE` IMO.

DHowett commented:
The one risk I am aware of is that we *will* delete user color schemes that ONLY change the cursor or selection colors.

I can add checks to avoid this, but the risk seems fairly low overall.

lhecker commented:
I think that's an acceptable risk.

DHowett commented:
@lhecker any blockers left?

DHowett commented:
Body updates

DHowett commented:
So! Because `_addOrMerge` is called during settings loading and we don't have a place to store "I did fixups" _outside of `FixupUserSettings`_, I opted to instead track the observable effects of merging.

[This line in CascadiaSettingsSerialization](https://github.com/microsoft/terminal/pull/12800/files#diff-00c4cab1589ce4b52be3a6d069d917bcf01e6983ecc312f0befaa1c6cd6f6efcR469-R470) ensures that we write out the settings _if_ we renamed a user color scheme. Because renaming adds a remapping entry, the condition will always reflect whether there have been any renames.

The one case this *does not* cover is when a user had an identical color scheme which we deleted, but no other settings changes. Honestly, I don't mind that terribly. If they go save their settings again it will disappear, and it is not harming anyone much by being there.

I can plumb through a "made changes" flag if we want to force a flush to disk when we _delete_ schemes.

PankajBhojwani commented on src/cascadia/TerminalSettingsEditor/ColorSchemeViewModel.cpp:
```diff
@@ -74,6 +74,11 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         _NotifyChanges(L"IsDefaultScheme");
     }
 
+    bool ColorSchemeViewModel::IsInBoxScheme() const
+    {
+        return _scheme.Origin() != Model::OriginTag::User;
```
nit: the function name and body are a little confusing since `OriginTag::InBox` exists, so from the function name I'd expect the body to just be `_scheme.Origin() == OriginTag::InBox`. Maybe rename the function to `IsNotUserScheme`?

DHowett replied:
Good point. I flipped the whole thing around to be `IsEditable`

PankajBhojwani commented on src/cascadia/TerminalSettingsEditor/ColorSchemesPageViewModel.cpp:
```diff
@@ -205,6 +192,24 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+                _AllColorSchemes.Append(schemeVM);
+                _viewModelToSchemeMap.Insert(schemeVM, scheme);
+                CurrentScheme(schemeVM);
+                CurrentPage(ColorSchemesSubPage::Base);
```
We're switching to the base page even though the next line immediately switches to the edit page, I believe this is because of the `Breadcrumb` logic in `MainPage`? Unrelated to this PR I should probably do something about that, this is ugh.

DHowett replied:
YEP that's exactly why. It can't update the page inline.

PankajBhojwani commented on src/cascadia/TerminalSettingsModel/ISettingsModelObject.idl:
```diff
@@ -0,0 +1,20 @@
+
+namespace Microsoft.Terminal.Settings.Model
+{
+    // This tag is used to identify the context in which the Profile was created
```
nit: remove "Profile" from this comment

DHowett replied:
Done
